src/lib.rs

```diff
@@ -55,7 +61,63 @@ where
             Error::Spi(spi) => write!(f, "Error::Spi({:?})", spi),
             Error::Gpio(gpio) => write!(f, "Error::Gpio({:?})", gpio),
             Error::UnexpectedStatus => f.write_str("Error::UnexpectedStatus"),
+            Error::SectorLength => f.write_str("Error::SectorLength"),
             Error::__NonExhaustive(_) => unreachable!(),
         }
     }
 }
+
+/// A trait for reading operations from a memory chip
+pub trait Read<Addr, SPI: Transfer<u8>, CS: OutputPin> {
+    /// Reads bytes from a memory chip
+    ///
+    /// # Parameters
+    /// * `addr`: The address to start reading at
+    /// * `buf`: The buffer to read buf.len() bytes into
+    fn read(&mut self, addr: Addr, buf: &mut [u8]) -> Result<(), Error<SPI, CS>>;
+}
+
+/// A trait for writing and erasing operations on a memory chip
+pub trait BlockDevice<Addr: TryInto<usize> + Copy, SPI: Transfer<u8>, CS: OutputPin>
+where
+    <Addr as core::convert::TryInto<usize>>::Error: core::fmt::Debug,
+{
+    /// The sector length in bytes, should be set to 1 for EEPROM implementations
+    const SECTOR_LENGTH: usize;
+
+    /// Erases bytes from the memory chip
+    ///
+    /// This function will return a `SectorLength` error if `amount` is not a multiple
+    /// of [SECTOR_LENGTH](BlockDevice::SECTOR_LENGTH)
+    ///
+    /// # Parameters
+    /// * `addr`: The address to start erasing at
+    /// * `amount`: The amount of bytes to erase, starting at `addr`
+    fn erase_bytes(&mut self, addr: Addr, amount: usize) -> Result<(), Error<SPI, CS>> {
```

Comment:
I think it would be better to have erasure work on a sector granularity instead of taking the amount of bytes here. This would make `erase_bytes_unchecked` unnecessary, too.

Reply:
It would also allow getting rid of the `TryInto<usize>` requirement, I believe.

Reply:
But on for example EEPROM chips you can also just erase a single byte of memory if you want to so that'd be dozens and dozens of calls do this function instead of just a simple for loop inside it.

Reply:
Oh, it should still take the number of sectors to erase I think. I just want to remove the need for the unsafe function and this wrapper.